mrmustard/physics/ansatze.py

```diff
@@ -0,0 +1,364 @@
+                to_keep.append(d)
+                d0 = d
+                mat, vec = self.mat[d0], self.vec[d0]
+        self.mat = math.gather(self.mat, to_keep, axis=0)
+        self.vec = math.gather(self.vec, to_keep, axis=0)
+        self.array = math.gather(self.array, to_keep, axis=0)
+        self._simplified = True
+
+    def _order_batch(self):
+        generators = [
+            itertools.chain(
+                math.asnumpy(self.vec[i]).flat,
+                math.asnumpy(self.mat[i]).flat,
+                math.asnumpy(self.array[i]).flat,
+            )
+            for i in range(self.batch_size)
+        ]
+        sorted_indices = argsort_gen(generators)
+        self.mat = math.gather(self.mat, sorted_indices, axis=0)
+        self.vec = math.gather(self.vec, sorted_indices, axis=0)
+        self.array = math.gather(self.array, sorted_indices, axis=0)
+
+
+class PolyExpAnsatz(PolyExpBase):
+    r"""
+    Represents the ansatz function:
+
+        :math:`F(z) = sum_i poly_i(z) exp(z^T A_i z / 2 + z^T b_i)`
+
+    where each :math:`poly_i` is a polynomial in ``z`` that can be expressed as:
+
+        :math:`poly_i(z) = sum_k c^(i)_k z^k`,
+
+    with ``k`` being a multi-index. The batch of arrays :math:`c^{(i)}` are not just array values but can be polynomials
+    of varying order, defined by the terms :math:`arr_k z^k` for each ``i``.
+    The matrices :math:`A_i` and vectors :math:`b_i` are parameters of the
+    exponential terms in the ansatz, and :math:`z` is a vector of variables.
+
+    Attributes:
+        A (Batch[Matrix]): The list of square matrices A_i
+        b (Batch[Vector]): The list of vectors b_i
+        c (Batch[Tensor]): The array of coefficients for the polynomial terms in the ansatz.
+
+    Example:
+        >>> A = [np.array([[1.0, 0.0], [0.0, 1.0]])]
+        >>> b = [np.array([1.0, 1.0])]
+        >>> c = [np.array(1.0)]
+        >>> F = PolyExpAnsatz(A, b, c)
+        >>> z = np.array([1.0, 2.0])
+        >>> print(F(z))  # prints the value of F at z
+    """
+
+    def __init__(self, A: Optional[Batch[Matrix]] = None, b: Optional[Batch[Vector]] = None, c: Batch[Tensor | Scalar] = [1.0], name: str = ""):
+        self.name = name
+        assert A is not None or b is not None, "Please provide either A or b."
+        dim = b[0].shape[-1] if A is None else A[0].shape[-1]
+        A = A if A is not None else np.zeros((len(b), dim, dim), dtype=b[0].dtype)
+        b = b if b is not None else np.zeros((len(A), dim), dtype=A[0].dtype)
+        super().__init__(mat=A, vec=b, array=c)
+
+    @property
+    def A(self) -> Batch[ComplexMatrix]:
+        return self.mat
+
+    @property
+    def b(self) -> Batch[ComplexVector]:
+        return self.vec
+
+    @property
+    def c(self) -> Batch[ComplexTensor]:
+        return self.array
+
+    @property
+    def degree(self) -> int:
+        if self.array.ndim == 1:
+            return 0
+        return self.array.shape[-1] - 1
+
+    def plot(
```

Comment:
Please add docstring here to describe the args, as I'm not sure I understand all of them